backend/src/bundles/lms-data/lms-data.service.ts

```diff
@@ -0,0 +1,135 @@
+import { type LMSDataRepository } from './lms-data.repository.js';
+import {
+    type LMSDataServerResponseDto,
+    type UserLMSDataDto,
+} from './types/types.js';
+
+class LMSDataService implements Service {
+    private lmsDataRepository: LMSDataRepository;
+    private requestsToLMSHeaders: { 'X-Token': string | undefined };
+
+    public constructor(
+        lmsDataRepository: LMSDataRepository,
+        token: string | undefined,
+    ) {
+        this.lmsDataRepository = lmsDataRepository;
+        this.requestsToLMSHeaders = { 'X-Token': token };
+    }
+
+    public async findByUserId(
+        userId: string,
+    ): Promise<UserLMSDataDto | undefined> {
+        const dataFromDB = await this.lmsDataRepository.findByUserId({
+            userId,
+        });
+        if (dataFromDB) {
+            return dataFromDB.toObject();
+        }
+
+        const user = await userRepository.findById(userId);
+        const userEmail = user?.toObject().email;
+
+        if (userEmail) {
+            const dataFromLMS = await this.findByUserEmailOnLMSServer(
+                userEmail,
+            );
+            if (dataFromLMS) {
+                const parsedLMSData = this.parseLMSServerData(
+                    userId,
+                    dataFromLMS,
+                );
+
+                const newDBRecord = await this.lmsDataRepository.create(
+                    LMSDataEntity.initialize(parsedLMSData),
+                );
+                return newDBRecord.toObject();
+            }
+
+            return undefined;
+        }
+
+        return undefined;
+    }
+
+    private async findByUserEmailOnLMSServer(
+        email: string,
+    ): Promise<LMSDataServerResponseDto | undefined> {
+        // TODO: specify type of return value
+
+        const url = new URL(LMSDataApiPath.LMS_SERVER);
+        url.searchParams.append('email', email);
+
+        const response = await http.load(url, {
+            headers: this.requestsToLMSHeaders,
+        });
+
+        //TODO: need change it to something proper
+        const data = (await response.json()) as LMSDataServerResponseDto;
+
+        // TODO: need to refactor this check on error, this was quick solution
+        if (!Object.keys(data).includes('talent')) {
+            return undefined;
+        }
+
+        return data;
+    }
+
+    // only for test, should be removed
+    public async testLMSServer(
+        email: string,
+    ): Promise<UserLMSDataDto | unknown> {
+        const url = new URL(LMSDataApiPath.LMS_SERVER);
+        url.searchParams.append('email', email);
+
+        const response = await http.load(url, {
+            headers: this.requestsToLMSHeaders,
+        });
+
+        const json = (await response.json()) as LMSDataServerResponseDto;
+        return this.parseLMSServerData(email, json);
+    }
+
+    private parseLMSServerData(
+        userId: string,
+        serverData: LMSDataServerResponseDto,
+    ): UserLMSDataDto {
+        return {
```

Comment:
It would be better to create a function for this (helper)
```lmsServerDataToLmsUserDataDTO()```

Reply:
name is too complicated. parse is widespread term. yes, in common logic of project it should be in helpers, but in that case I think, to better understanding let leave it in service

Reply:
I have similar thing, and I placed it in helpers.
I think that service should be clear to read.

Reply:
okay then